CIP-0021/CIP-0021.md

```diff
@@ -51,6 +51,14 @@ While the Cardano CDDL specification usually does not limit the size of integers
 
 The numbers of inputs, outputs, certificates, withdrawals, tokens in an asset group, and other similar transaction elements must not exceed `UINT16_MAX`, i.e. 65535.
 
+**Optional empty lists and maps**
+
+Unless mentioned otherwise in this CIP, optional empty lists and maps must not be included as part of the transaction body or its elements.
+
+**Outputs**
+
+Outputs containing no multiassets must be serialized as a simple tuple. I.e. `[address, coin]` instead of the alternative `[address, [coin, {}]]`.
```

Comment:
It seems to me we support empty asset groups --- should it mentioned in this document? Perhaps not, but it's a natural question to ask and it's not obvious if it falls under "optional empty lists and maps".

Reply:
Trezor doesn't allow empty asset groups. I see nothing restricting this on Ledger. Tokens are shown on Trezor only as a combination of policyId+assetName, thus I don't know what should be shown if an empty asset group is provided - is it shown on Ledger? If so how? I might check myself tomorrow.

To clarify the CIP though, we could add "If a map's key is also an empty list or an empty map it is also not to be included unless stated otherwise." to "Optional empty lists and maps".

Reply:
Hm, it seems to me that Ledger actually crashes (restarts) when signing a transaction with the following output with empty `tokens` in `tokenBundle`:
```
{
  destination: {
    type: TxOutputDestinationType.THIRD_PARTY,
    params: {
      addressHex:
        "82d818582183581c9e1c71de652ec8b85fec296f0685ca3988781c94a2e1a5d89d92f45fa0001a0d0c2561",
    },
  },
  amount: "3003112",
  tokenBundle: [
    {
      policyIdHex:
        "95a292ffee938be03e9bae5657982a74e9014eb4960108c9e23a5b39",
      tokens: [],
    },
  ],
},
```
When I remove the `tokenBundle` or add tokens to `tokens` it works.

Reply:
I just realised we might be mixing up `tokenBundle` and `asset group` - I'm not sure which is which right now. Sending an empty `tokenBundle: []` produces the same tx hash as not sending the `tokenBundle` does.